From 790b109854ddb0a208c5c0ce6af74d19e033779e Mon Sep 17 00:00:00 2001
From: Sarath Babu Naidu Gaddam <sarath.babu.naidu.gaddam@xilinx.com>
Date: Thu, 16 Jun 2022 07:13:13 -0600
Subject: [PATCH 7/7] ptp: clockmatrix: Add defer probe

When board is in SD bootmode, IDT driver(clock matrix) loads first
then rootfs so idt driver will not see idtcm.bin at first time so
added defer probe.

Signed-off-by: Sarath Babu Naidu Gaddam <sarath.babu.naidu.gaddam@xilinx.com>
---
 drivers/ptp/ptp_clockmatrix.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/ptp/ptp_clockmatrix.c b/drivers/ptp/ptp_clockmatrix.c
index c9d451b..c45c487 100644
--- a/drivers/ptp/ptp_clockmatrix.c
+++ b/drivers/ptp/ptp_clockmatrix.c
@@ -2424,9 +2424,14 @@ static int idtcm_probe(struct platform_device *pdev)
 
 	err = idtcm_load_firmware(idtcm, &pdev->dev);
 
-	if (err)
+	if (err) {
 		dev_warn(idtcm->dev, "loading firmware failed with %d", err);
-
+		if( err == -ENOENT) {
+			mutex_unlock(idtcm->lock);
+			dev_warn(idtcm->dev,"deferring the probe\n");
+			return -EPROBE_DEFER;
+		}
+	}
 	wait_for_chip_ready(idtcm);
 
 	if (idtcm->tod_mask) {
-- 
2.7.4

